agent/service.go

```diff
@@ -178,6 +180,8 @@ func New(ctx context.Context, logger *slog.Logger, eventSvc events.Service, cmp
 }
 
 func (as *agentService) Algo(ctx context.Context, algo Algorithm) error {
+	as.mu.Lock()
+	defer as.mu.Unlock()
```

Comment:
Can you move this lock closer to where it is used, i.e. wehere you access (and set) algorithm.